src/humanloop/decorators/tool.py

```diff
@@ -0,0 +1,494 @@
+                        "value": {"type": _JSON_SCHEMA_ANY},
+                    },
+                }
+            else:
+                arg_type = {
+                    "type": "object",
+                    "properties": {
+                        "key": {"type": _JSON_SCHEMA_ANY},
+                        "value": {"type": _JSON_SCHEMA_ANY},
+                    },
+                }
+        else:
+            arg_type = {
+                "type": "object",
+                "properties": {
+                    "key": _annotation_parse_to_json_schema(arg.key_annotation),  # type: ignore
+                    "value": _annotation_parse_to_json_schema(arg.value_annotation),  # type: ignore
+                },
+            }
+
+    elif isinstance(arg, _ParsedPrimitiveAnnotation):
+        if arg.annotation is builtins.str:
+            arg_type = {"type": "string"}
+        if arg.annotation is builtins.int:
+            arg_type = {"type": "integer"}
+        if arg.annotation is builtins.float:
+            arg_type = {"type": "number"}
+        if arg.annotation is builtins.bool:
+            arg_type = {"type": "boolean"}
+        if arg.annotation is Parameter.empty:
+            # JSON Schema dropped support for 'any' type, we allow any type as a workaround
+            arg_type = {"type": _JSON_SCHEMA_ANY}
+
+    else:
+        raise ValueError(f"Unsupported annotation type: {arg}")
+
+    if is_optional:
+        if isinstance(arg, _ParsedUnionAnnotation):
+            for type_option in arg_type["anyOf"]:
+                if (
+                    isinstance(type_option["type"], list)  # type: ignore
+                    and "null" not in type_option["type"]  # type: ignore
+                ):  # type: ignore
+                    type_option["type"] = [*type_option["type"], "null"]  # type: ignore
+                elif not isinstance(type_option["type"], list):  # type: ignore
+                    type_option["type"] = [type_option["type"], "null"]  # type: ignore
+        else:
+            if isinstance(arg_type["type"], list) and "null" not in arg_type["type"]:  # type: ignore
+                arg_type = {**arg_type, "type": [*arg_type["type"], "null"]}  # type: ignore
+            elif not isinstance(arg_type["type"], list):  # type: ignore
+                arg_type = {**arg_type, "type": [arg_type["type"], "null"]}  # type: ignore
+
+    return arg_type
+
+
+def _parameter_is_optional(
+    parameter: inspect.Parameter,
+) -> bool:
+    """Check if tool parameter is mandatory.
+
+    Examples:
+        Optional[T] -> True
+        T | None -> True
+        T -> False
+    """
+    # Check if the parameter can be None, either via Optional[T] or T | None type hint
+    origin = typing.get_origin(parameter.annotation)
+    # sub_types refers to T inside the annotation
+    sub_types = typing.get_args(parameter.annotation)
+    return origin is typing.Union and len(sub_types) > 0 and sub_types[-1] is type(None)
```

Comment:
does this work for `None | T`

testing locally, also running problems with `origin is typing.Union` - `origin` is `types.UnionType` for me.
(python 3.11)

Reply:
I have added support for parsing python 3.10 style union annotation, added tests for it, and created a new action that runs the test suite on both 3.9 and 3.12